Updates to Licensing & README.md

Rustymage opened the pull request.

Rustymage committed:
License Removal

Rustymage committed:
License Update

MIT License Added

Rustymage committed:
READ.md updated

Installation instructions added

Rustymage committed:
READ.md Updates

osx_services & font titles added

Rustymage committed:
README.md Updated

Font section created and updated

Rustymage committed:
README.md

Bullet point added

Rustymage committed:
README.md Updated

osx_services updated

Rustymage committed:
Update README.md

Link added to Robot font and removal of Avenir (licensed font)

Rustymage committed:
Font Removal

Removal of Avenir font as it's a paid font.

Rustymage committed:
Update to README.md

Link fixed

Rustymage committed:
Update README.md

stevequinn commented on README.md:
````diff
@@ -23,13 +37,29 @@ options:
+
+```photoborder/fonts```
+
+Should you wish to use another font you will need to manually update ```text.py```
````
While I love the detail on how fonts can be updated, can I suggest we move the font name strings ""Roboto-Regular.ttf" and "Roboto-Regular.ttf" to a .env file so the end user is not modifying code directly? This would result in:

1. A new dot.env file in the root that is committed to git that has the example .env structure:

FONTNAME="Roboto-Regular.ttf"
BOLDFONTNAME="Roboto-Medium.ttf"

2. A description in the readme installation section that tells the user they need to copy the example dot.env to .env.

3. An updated .gitignore with .env so it is not committed to git. 

4. A font section in the readme like you have that describes updating the .env vars if the user wants to change the font.

5. An updated text.py file that leverages the python-dotenv package (this will also need to be added to requirements.txt), to load the env vars. See below for a code example of this. 

```python
# !/usr/bin/env python

"""
Environment variable loader
"""

import os
from dotenv import load_dotenv

IS_LOADED = False

if not IS_LOADED:
    load_dotenv()  # take environment variables from .env.
    IS_LOADED = True

def get(name: str) -> str:
    """
    Get the value of an environment variable

    Args:
        name (str): The env name

    Returns:
        str: env value
    """
    return os.getenv(name)
```
    
    If you don't have experience with env vars I'd be happy to implement this, or if you have any aversion to the idea let me know :)
    
    Thanks.

stevequinn replied:
Alternatively, now that I think of it, it could be moved to a parameter with appropriate bounds checking to reduce complexity:

parser.add_argument('-f', '--font', default='Roboto-Regular.ttf', help='Font file in fonts directory')
parser.add_argument('-fb', '--fontbold', default='Roboto-Medium.ttf', help='Bold font file in fonts directory')

Rustymage replied:
Let me take a look. I did think of an environment file but with time and the need to research how to do it (my coding experience is limited, but I enjoy learning) I was going to resolve at a later date. Although, I'd be happy to pull that timeline forward and hold this PR open for a bit.

I'll have a go and see how it goes based on the guidance above. 

👍

stevequinn replied:
I almost think parameters might be a better option? It's less initial configuration for the user, and easy to pass in new options and have sane defaults. 

What do you think?

Rustymage replied:
So I've created it as you suggested (feels the best method) and added a validation font function to gracefully exit if the font is not available in the directory. 

I did utilise **CoPilot** for support as I got stuck on passing arguments to the main function and had it incorrectly defined. I had global definitions for font/bold font in `text.py` and that clearly wasn't working. Shifting to `main.py` under the main function fixed that with the parameters modified in `border.py`,` text.py `& `main.py`

```
font_path = os.path.join(FONTDIR, args.font)
bold_font_path = os.path.join(FONTDIR, args.fontbold)
```
Figured it out with help. The `README.md` has been updated and a small tweak to `.gitignore` too.

I won't be offended if it needs to be tidied. However, it does work as intended.

👍

stevequinn replied:
Good stuff, thanks for this. I just streamlined the error reporting a little bit as there were some duplicates, and removed some redundant code. I also added python logging so we can use that instead of print statements from now on. 

I'll wait for you to review before I merge into main.

stevequinn commented on LICENSE:
```diff
@@ -1,24 +1,7 @@
-This is free and unencumbered software released into the public domain.
```
Can you add the licence type header here just so it's clear? As per https://github.com/ivandokov/phockup/blob/master/license

The MIT License (MIT)

Rustymage replied:
Good spot, with Github picking up the license type and adding to a header I missed this. Will resolve.

Rustymage committed:
Resolving Comment on PR

Resolving https://github.com/stevequinn/photoborder/pull/8#discussion_r1841149656

Rustymage committed:
Font Arg Parse

Rustymage committed:
PR 8 Update

Utilised CoPilot to add argument for font name changes

Rustymage committed:
PR8 Font Validation

Validate font using CoPilot help

Rustymage committed:
Update to files

.gitignore & READ.md update

Rustymage committed:
Merge pull request #1 from Rustymage/Dev

Dev to Main

Rustymage committed:
Remove Test Image

Rustymage committed:
Comments

Re-added comments deleted

Rustymage committed:
.

Rustymage committed:
Validation Check

Improvement in check

Rustymage committed:
Validation Check

Correction to duplication of error messages.

Rustymage commented on text.py:
```diff
@@ -4,42 +4,48 @@
-# FONTINDEX = 8
-FONTNAME = os.path.join(FONTDIR, "Roboto-Regular.ttf")
-BOLDFONTNAME = os.path.join(FONTDIR, "Roboto-Medium.ttf")
 FONTINDEX = 0
```
Not sure `FONTINDEX` is needed, except where a single file contains multiple styles. Is this still necessary? When adding both bold and regular font files, this feels superfluous?

Rustymage commented on main.py:
```diff
@@ -131,6 +118,18 @@ def main():
     font_path = os.path.join(FONTDIR, args.font)
     bold_font_path = os.path.join(FONTDIR, args.fontbold)
 
+    # Validate fonts before processing any images
```
Validation check of font directory if argument for font is utilised.

Rustymage commented on main.py:
```diff
@@ -125,9 +125,7 @@ def main():
-    if error_messages:
-        for message in error_messages:
-            print(message)
+    if not os.path.isfile(font_path) or not os.path.isfile(bold_font_path):
```
This spits out an error depending on which font is missing when called by the argument and then gracefully exits.

stevequinn committed:
- Streamlines error reporting
- Removes redundant code

stevequinn committed:
- Adds python logging instead of print statements.

Rustymage commented on border.py:
```diff
@@ -123,13 +123,13 @@ def draw_border(img: Image, border: Border) -> Image:
+    font_size = tm.get_optimal_font_size("Test string", border.bottom * multiplier, fontpath)
+    heading_font_size = tm.get_optimal_font_size("Test string", border.bottom * (multiplier + 0.02), boldfontpath)
+    font = tm.create_font(font_size, fontpath)
+    heading_font = tm.create_font(heading_font_size, boldfontpath)
```
I see what you've done here - reduced the number of functions. Neat.

Rustymage commented on main.py:
```diff
@@ -49,8 +51,8 @@ def process_image(path: str, add_exif: bool, add_palette: bool, border_type: Bor
         border_type (BorderType): The type of border to add to the photo.
-        fontname (str): Path to the font file
-        boldfontname (str): Path to the bold font file
+        fontname (str): name of the font file to use
```
Agreed - I relied a little too much on CoPilot naming and wasn't consistent!

Rustymage commented on main.py:
```diff
@@ -70,7 +72,18 @@ def process_image(path: str, add_exif: bool, add_palette: bool, border_type: Bor
         exif = get_exif(img)
         if exif:
-            img_with_border = draw_exif(img_with_border, exif, border, fontname, boldfontname)
+            moduledir = os.path.dirname(os.path.abspath(__file__))
```
Can you help me understand why this is outside the main function? When I did that, it threw errors but it may have been a result of my parameters eleswhere...

stevequinn replied:
I moved it to the exif if statement as we only want to do that processing if the user has requested exif to be added, and there is actual exif data to use. 

I'm not sure why it wouldn't work for you though, as the __file__ abspath should be resolved from within any function. 

This function is turning into a bit of spaghetti code though so at some point I'll have a go at refactoring it.

Rustymage commented on main.py:
```diff
@@ -33,12 +34,13 @@ def parse_arguments():
     return parser.parse_args()
 
+
 def process_image(path: str, add_exif: bool, add_palette: bool, border_type: BorderType, fontname: str, boldfontname: str) -> str:
```
Should this be `fontpath` & `boldfontpath` now?

stevequinn replied:
So, it's still the font name at this  point right? The process_image function then determines the paths within the exif if statement.

Rustymage commented on text.py:
```diff
@@ -4,46 +4,28 @@
-    font = ImageFont.truetype(fontname, size)
-    return font
-
-def create_bold_font(size: int, fontname: str) -> ImageFont.FreeTypeFont:
```
I believe the removal of this is because we use the `create_font` function for both regular & bold fonts now?

Does this work if `font` and **`bold font`** are different styles?

stevequinn replied:
That is  a good point. I'll close this PR now as it is getting quite large and covering a few bases, then how about we create another PR to add support for a font index parameter which defaults to None or 0 (need to test), and gets passed through to this create_font function?

Rustymage reviewed (commented):
My review is mostly questions to educate myself and understand the improvements 👍
